### DISCUSSION TITLE
Add a new field without losing data

### DISCUSSION THREAD

**Discussion by muco-rolle** [prisma/prisma#8724] [Q&A]
I'm new to Prisma and SQL in general so I need help to understand this,  

so take for example I have a database with a table called  `posts` I want to add a required field like `description` I saw that I can run `npx prisma migrate dev --create-only` and do some changes in the migration file, but I do not understand what changes I can do in that file so that I'm able to add the new field and keep the data I have.

**Comment by janpio** [comment]
If you just want to add a field, it should not mess with your existing data at all.
What changes do you want to do in the migration file and why? What do you not like about the created SQL?

**Reply by muco-rolle** [reply]
I'm just getting started with Prisma I started with the following model

```prisma
model Post {
    id        Int      @id @default(autoincrement())
    title     String
    createdAt DateTime @default(now())
    updatedAt DateTime @updatedAt
}
```

Now I want to add on that model a `body` field which is required

```prisma
model Post {
    id        Int      @id @default(autoincrement())
    title     String
    body   String
    createdAt DateTime @default(now())
    updatedAt DateTime @updatedAt
}
```

When I run `npx prisma migrate dev` I get this error

```
Error: 
⚠️ We found changes that cannot be executed:

  • Step 0 Added the required column `body` to the `Post` table without a default value. There are 1 rows in this table, it is not possible to execute this step.

You can use npx prisma migrate dev --create-only to create the migration file, and manually modify it to address the underlying issue(s).
Then run npx prisma migrate dev to apply it and verify it works.
```

and then when I run `npx prisma migrate dev --create-only` I'm able to create a migration

 this is the content of the migration just created
 
 ```
 /*
  Warnings:

  - Added the required column `body` to the `Post` table without a default value. This is not possible if the table is not empty.

*/
-- AlterTable
ALTER TABLE "Post" ADD COLUMN     "body" TEXT NULL;
 ```
 
 with this message `and manually modify it to address the underlying issue(s).`  from the error above, I don't see what to do to be able to add that  field to `Post` table without losing data I already have.

**Comment by ryands17** [comment]
@muco-rolle 👋 

It's not possible to add a required field directly on a model when you have existing data. You need to add a new field as an optional one first, seed it with data and then make it required.

So to summarise:

1. Add the body field to the model like this: `body String?`
2. Seed the existing fields with the body. An empty string will also do (`''`)
3. Add another migration to make it required. Remove the `?` from the field and then run migrate.

**Reply by samyarkd** [reply]
` metadata               String?`
in my case it still requires the reset even though i added the optional property

**Reply by kishmiryan-karlen** [reply]
Or you can just declare it as a required field and specify a default value: `body String @default("")`

**Comment by residentcode** [comment]
I wish prisma become dynamic ORM like Django, keep going guys nothing is impossible.

**Reply by santoshtechwiz** [reply]
Yes agree with you I wrongly choose this tool. I have reset my database 3 times to add a column .I am looking for tools that can execute raw sql on the top of prisma

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
